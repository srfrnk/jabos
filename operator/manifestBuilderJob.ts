import { ApiObjectProps } from 'cdk8s';
import builderJob from './builderJob';
import { Container, KubeJobProps, Quantity, Volume } from './imports/k8s';
import { GitRepositoryPropsEx } from './misc';
import settings from './settings';

export default function (options: {
  object: ApiObjectProps,
  repo: GitRepositoryPropsEx,
  imagePrefix: string,
  buildNumber: string,
  type: string,
  containers: Container[],
  volumes: Volume[],
  metricName: string,
  metricLabels: { [key: string]: string; },
  kind: string,
  controller: string
}): KubeJobProps {
  const containers = (options.containers || []).map(container => {
    return {
      ...container,
      volumeMounts: [
        ...(container.volumeMounts || []),
        {
          name: "manifests",
          mountPath: "/manifests",
        }
      ]
    };
  });

  const targetNamespace = options.object.spec.targetNamespace;

  return builderJob({
    object: options.object,
    repo: options.repo,
    jobNamePrefix: 'manifest',
    imagePrefix: settings.imagePrefix(),
    buildNumber: settings.buildNumber(),
    serviceAccountName: `builder-${options.object.spec.gitRepository}`,
    type: options.type,
    commitLabel: "deployedCommit",
    metricName: options.metricName,
    metricLabels: options.metricLabels,
    labels: { type: "manifest-builder", "manifest-builder-type": options.type },
    containers: [
      ...containers,
      {
        image: `${options.imagePrefix}manifest-deployer:${options.buildNumber}`,
        args: [targetNamespace, options.type, options.object.metadata.name],
        stdin: true,
        tty: true,
        env: [
          {
            name: "NAMESPACE",
            value: options.object.metadata.namespace
          },
          {
            name: "TARGET_NAMESPACE",
            value: targetNamespace
          },
          {
            name: "TYPE",
            value: options.type
          },
          {
            name: "NAME",
            value: options.object.metadata.name
          },
          {
            name: "KIND",
            value: options.kind
          },
          {
            name: "CONTROLLER",
            value: options.controller
          },
        ],
        name: "manifest-deployer",
        volumeMounts: [
          {
            name: "manifests",
            mountPath: "/manifests",
            readOnly: true
          }
        ],
        resources: {
          limits: {
            cpu: Quantity.fromString("500m"),
            memory: Quantity.fromString("500Mi")
          },
          requests: {
            cpu: Quantity.fromString("100m"),
            memory: Quantity.fromString("100Mi")
          }
        },
      }
    ],
    volumes: [
      ...options.volumes,
      {
        name: "manifests",
        emptyDir: {}
      }
    ]
  });
}