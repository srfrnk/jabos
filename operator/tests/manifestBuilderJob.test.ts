import { App, Chart } from 'cdk8s';
import { Container, KubeJob, Volume } from '../imports/k8s';
import manifestBuilderJob from '../manifestBuilderJob';
import { clearMock, setMock } from './settingsMock';

beforeEach(() => {
  setMock();
});

afterEach(() => {
  clearMock();
});

test('manifestBuilderJob', () => {
  const options = {
    object: {
      apiVersion: 'apiVersion_value',
      kind: 'kind_version',
      metadata: {
        name: 'name_value',
        namespace: 'namespace_value',
        uid: 'uid_value',
      },
      spec: {
        targetNamespace: 'targetNamespace_value',
        gitRepository: 'gitRepository_value'
      },
    },
    repo: {
      spec: {
        url: 'repo_url',
        branch: 'repo_branch'
      },
      status: {
        latestCommit: 'latestCommit_value'
      }
    },
    imagePrefix: 'imagePrefix_value',
    buildNumber: 'buildNumber_value',
    type: 'type_value',
    containers: [{} as Container],
    volumes: [],
    metricName: 'metricName_value',
    metricLabels: {},
    kind: 'kind_value',
    controller: 'controller_name'
  }
  const chart = new Chart(new App(), 'chart');
  expect(new KubeJob(chart, 'job', manifestBuilderJob(options)).toJson()).toMatchSnapshot();
});

test('manifestBuilderJob no containers', () => {
  const options = {
    object: {
      apiVersion: 'apiVersion_value',
      kind: 'kind_version',
      metadata: {
        name: 'name_value',
        namespace: 'namespace_value',
        uid: 'uid_value',
      },
      spec: {
        targetNamespace: 'targetNamespace_value',
        gitRepository: 'gitRepository_value'
      },
    },
    repo: {
      spec: {
        url: 'repo_url',
        branch: 'repo_branch'
      },
      status: {
        latestCommit: 'latestCommit_value'
      }
    },
    imagePrefix: 'imagePrefix_value',
    buildNumber: 'buildNumber_value',
    type: 'type_value',
    containers: undefined,
    volumes: [],
    metricName: 'metricName_value',
    metricLabels: {},
    kind: 'kind_value',
    controller: 'controller_name'
  }
  const chart = new Chart(new App(), 'chart');
  expect(new KubeJob(chart, 'job', manifestBuilderJob(options)).toJson()).toMatchSnapshot();
})