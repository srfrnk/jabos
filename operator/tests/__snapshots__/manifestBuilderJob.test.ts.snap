// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manifestBuilderJob 1`] = `
Object {
  "apiVersion": "batch/v1",
  "kind": "Job",
  "metadata": Object {
    "labels": Object {
      "manifest-builder-type": "type_value",
      "type": "manifest-builder",
    },
    "name": "manifest-name_value-latestCommit_value",
  },
  "spec": Object {
    "activeDeadlineSeconds": 1234,
    "backoffLimit": 1234,
    "completionMode": "NonIndexed",
    "completions": 1,
    "parallelism": 1,
    "template": Object {
      "metadata": Object {
        "labels": Object {
          "builder": "manifest-name_value-latestCommit_value",
        },
        "name": "manifest-name_value-latestCommit_value",
      },
      "spec": Object {
        "containers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "TYPE",
                "value": "type_value",
              },
              Object {
                "name": "NAME",
                "value": "name_value",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "COMMIT_LABEL",
                "value": "deployedCommit",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "namespace_value",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "uid_value",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "settings_imagePrefixpost-builder:settings_build_number",
            "imagePullPolicy": "Always",
            "name": "post-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/timer",
                "name": "timer",
                "readOnly": true,
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
            ],
          },
        ],
        "initContainers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "URL",
                "value": "repo_url",
              },
              Object {
                "name": "BRANCH",
                "value": "repo_branch",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "namespace_value",
              },
              Object {
                "name": "NAME",
                "value": "name_value",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "uid_value",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "settings_imagePrefixpre-builder:settings_build_number",
            "imagePullPolicy": "Always",
            "name": "pre-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/gitTemp",
                "name": "git-temp",
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/timer",
                "name": "timer",
              },
            ],
          },
          Object {
            "env": Array [
              Object {
                "name": "NAMESPACE",
                "value": "namespace_value",
              },
              Object {
                "name": "NAME",
                "value": "name_value",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "uid_value",
              },
            ],
            "imagePullPolicy": "Always",
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/manifests",
                "name": "manifests",
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/build",
                "name": "build",
              },
            ],
          },
          Object {
            "args": Array [
              "targetNamespace_value",
              "type_value",
              "name_value",
            ],
            "env": Array [
              Object {
                "name": "NAMESPACE",
                "value": "namespace_value",
              },
              Object {
                "name": "TARGET_NAMESPACE",
                "value": "targetNamespace_value",
              },
              Object {
                "name": "TYPE",
                "value": "type_value",
              },
              Object {
                "name": "NAME",
                "value": "name_value",
              },
              Object {
                "name": "KIND",
                "value": "kind_value",
              },
              Object {
                "name": "CONTROLLER",
                "value": "controller_name",
              },
              Object {
                "name": "NAMESPACE",
                "value": "namespace_value",
              },
              Object {
                "name": "NAME",
                "value": "name_value",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "uid_value",
              },
            ],
            "image": "imagePrefix_valuemanifest-deployer:buildNumber_value",
            "imagePullPolicy": "Always",
            "name": "manifest-deployer",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "stdin": true,
            "tty": true,
            "volumeMounts": Array [
              Object {
                "mountPath": "/manifests",
                "name": "manifests",
                "readOnly": true,
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/build",
                "name": "build",
              },
            ],
          },
        ],
        "restartPolicy": "Never",
        "securityContext": Object {
          "runAsNonRoot": true,
        },
        "serviceAccountName": "builder-gitRepository_value",
        "volumes": Array [
          Object {
            "emptyDir": Object {},
            "name": "manifests",
          },
          Object {
            "emptyDir": Object {},
            "name": "git-temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "build",
          },
          Object {
            "emptyDir": Object {},
            "name": "timer",
          },
        ],
      },
    },
    "ttlSecondsAfterFinished": 1234,
  },
}
`;

exports[`manifestBuilderJob no containers 1`] = `
Object {
  "apiVersion": "batch/v1",
  "kind": "Job",
  "metadata": Object {
    "labels": Object {
      "manifest-builder-type": "type_value",
      "type": "manifest-builder",
    },
    "name": "manifest-name_value-latestCommit_value",
  },
  "spec": Object {
    "activeDeadlineSeconds": 1234,
    "backoffLimit": 1234,
    "completionMode": "NonIndexed",
    "completions": 1,
    "parallelism": 1,
    "template": Object {
      "metadata": Object {
        "labels": Object {
          "builder": "manifest-name_value-latestCommit_value",
        },
        "name": "manifest-name_value-latestCommit_value",
      },
      "spec": Object {
        "containers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "TYPE",
                "value": "type_value",
              },
              Object {
                "name": "NAME",
                "value": "name_value",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "COMMIT_LABEL",
                "value": "deployedCommit",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "namespace_value",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "uid_value",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "settings_imagePrefixpost-builder:settings_build_number",
            "imagePullPolicy": "Always",
            "name": "post-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/timer",
                "name": "timer",
                "readOnly": true,
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
            ],
          },
        ],
        "initContainers": Array [
          Object {
            "args": Array [],
            "env": Array [
              Object {
                "name": "URL",
                "value": "repo_url",
              },
              Object {
                "name": "BRANCH",
                "value": "repo_branch",
              },
              Object {
                "name": "COMMIT",
                "value": "latestCommit_value",
              },
              Object {
                "name": "METRIC_NAME",
                "value": "metricName_value",
              },
              Object {
                "name": "METRIC_LABELS",
                "value": "{}",
              },
              Object {
                "name": "NAMESPACE",
                "value": "namespace_value",
              },
              Object {
                "name": "NAME",
                "value": "name_value",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "uid_value",
              },
              Object {
                "name": "JABOS_OPERATOR_URL",
                "value": "http://operator.settings_namespace:1234/",
              },
            ],
            "image": "settings_imagePrefixpre-builder:settings_build_number",
            "imagePullPolicy": "Always",
            "name": "pre-builder",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "volumeMounts": Array [
              Object {
                "mountPath": "/gitTemp",
                "name": "git-temp",
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/timer",
                "name": "timer",
              },
            ],
          },
          Object {
            "args": Array [
              "targetNamespace_value",
              "type_value",
              "name_value",
            ],
            "env": Array [
              Object {
                "name": "NAMESPACE",
                "value": "namespace_value",
              },
              Object {
                "name": "TARGET_NAMESPACE",
                "value": "targetNamespace_value",
              },
              Object {
                "name": "TYPE",
                "value": "type_value",
              },
              Object {
                "name": "NAME",
                "value": "name_value",
              },
              Object {
                "name": "KIND",
                "value": "kind_value",
              },
              Object {
                "name": "CONTROLLER",
                "value": "controller_name",
              },
              Object {
                "name": "NAMESPACE",
                "value": "namespace_value",
              },
              Object {
                "name": "NAME",
                "value": "name_value",
              },
              Object {
                "name": "OBJECT_UID",
                "value": "uid_value",
              },
            ],
            "image": "imagePrefix_valuemanifest-deployer:buildNumber_value",
            "imagePullPolicy": "Always",
            "name": "manifest-deployer",
            "resources": Object {
              "limits": Object {
                "cpu": "500m",
                "memory": "500Mi",
              },
              "requests": Object {
                "cpu": "100m",
                "memory": "100Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                ],
              },
              "readOnlyRootFilesystem": true,
              "runAsNonRoot": true,
            },
            "stdin": true,
            "tty": true,
            "volumeMounts": Array [
              Object {
                "mountPath": "/manifests",
                "name": "manifests",
                "readOnly": true,
              },
              Object {
                "mountPath": "/tmp",
                "name": "temp",
              },
              Object {
                "mountPath": "/build",
                "name": "build",
              },
            ],
          },
        ],
        "restartPolicy": "Never",
        "securityContext": Object {
          "runAsNonRoot": true,
        },
        "serviceAccountName": "builder-gitRepository_value",
        "volumes": Array [
          Object {
            "emptyDir": Object {},
            "name": "manifests",
          },
          Object {
            "emptyDir": Object {},
            "name": "git-temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "temp",
          },
          Object {
            "emptyDir": Object {},
            "name": "build",
          },
          Object {
            "emptyDir": Object {},
            "name": "timer",
          },
        ],
      },
    },
    "ttlSecondsAfterFinished": 1234,
  },
}
`;
